beacon-chain/blockchain/stategenerator/state_generator.go

```diff
@@ -7,37 +7,46 @@ import (
 	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
 	"github.com/prysmaticlabs/prysm/beacon-chain/db"
 	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
 	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/sirupsen/logrus"
 )
 
-// GenerateStateFromSlot generates state from the last finalized epoch till the specified slot.
-func GenerateStateFromSlot(ctx context.Context, db *db.BeaconDB, slot uint64) (*pb.BeaconState, error) {
-	fState, err := db.FinalizedState()
+// GenerateStateFromBlock generates state from the last finalized epoch till the specified block.
+func GenerateStateFromBlock(ctx context.Context, db *db.BeaconDB, block *pb.BeaconBlock) (*pb.BeaconState, error) {
+	fState, err := db.HistoricalStateFromSlot(block.Slot)
 	if err != nil {
 		return nil, err
 	}
 
-	if fState.Slot > slot {
+	if fState.Slot > block.Slot {
 		return nil, fmt.Errorf(
 			"requested slot %d < current slot %d in the finalized beacon state",
```

Comment:
is this check still needed? if yes maybe use historical state rather than finalized beacon state

Reply:
Not sure, but I am going to leave it here, just in case.

Reply:
yeah we cant generate a new state, if the state we are using has a higher slot than the received block. The naming is off though, will get it changed